apps/core/declarativewebcontainer.cpp

```diff
@@ -528,13 +529,35 @@ int DeclarativeWebContainer::activateTab(int tabId, const QString &url)
 int DeclarativeWebContainer::requestTabWithOwner(int tabId, const QString &url, uint ownerPid)
 {
     bool activated = m_model->activateTabById(tabId);
-    if (!activated) {
-        tabId = m_model->newTab(url);
-        if (ownerPid) {
-            m_tabOwners.insert(tabId, ownerPid);
+    // We should only create or activate tabs once the model has loaded
+    if (m_model->loaded()) {
+        // The tab model has already loaded, so we can go ahead and create the tab
+        if (!activated) {
+            tabId = m_model->newTab(url);
+            if (ownerPid) {
+                m_tabOwners.insert(tabId, ownerPid);
+            }
+        } else {
+            load(url, true);
         }
     } else {
-        load(url, true);
+        // The model has yet to loaded, so queue creation of the tab
+        QMetaObject::Connection * const connection = new QMetaObject::Connection;
+        *connection = connect(m_model.data(), &DeclarativeTabModel::loadedChanged, [this, tabId, url, ownerPid, activated, connection]() {
+            // We assume that m_model->loaded() is now set to true
+            int activatedTab = requestTabWithOwner(tabId, url, ownerPid);
+            qCDebug(lcCoreLog) << "Delaying tab request created tabId:" << activatedTab;
+            m_reservedTabs = 0;
+            // Single-shot connection
+            QObject::disconnect(*connection);
+            delete connection;
+        });
+
+        // Return the predicted tabId
```

Comment:
I'm just wondering which would make more sense: predicting `tabId` or delaying D-Bus call return until it's known.

Reply:
Thanks for the suggestion. I initially tried delaying but hit some difficulties, but I've had another go and I think this is nicer now. See the additional commit; what do you think?